creator-node/src/services/stateMachineManager/QueueInterfacer.js

```diff
@@ -0,0 +1,28 @@
+/**
+ * Singleton for interfacing with all the queues. Job processors must be functions
+ * that take simple, serializable types, so job processors that need to enqueue other jobs
+ * will use this interfacer instead of being passed a non-serializable queue.
+ */
```

Comment:
bump @theoilie

Reply:
read through how Bull spawns new processes [here](https://github.com/OptimalBits/bull/blob/edfbd163991c212dd6548875c22f2745f897ae28/lib/process/child-pool.js#L108-L123). it's using Node's `child_process.fork()`, which means it can send messages but their scope completely changes ([example SO thread](https://stackoverflow.com/questions/61304475/node-js-sending-an-object-from-a-child-process)) so it really only works for raw data and not something like libs/sdk where we need the child process to reflect state changes that happen. so unfortunately it looks like we do have to keep it completely independent and re-initialize libs

following up from offline sync, we should leave the jobs that require shared state on the main thread and incrementally move them over to their own processes as possible. this is the shared state each job currently needs:

**updateReplicaSet job processor:**
- `UserReplicaSetManagerClient.updateReplicaSet()` (libs)
- `ServiceProvider.autoSelectCreatorNodes()` (libs)
- `CNodeToSpIdMapManager.getCNodeEndpointToSpIdMap()` (state machine)
- `retrieveClockValueForUserFromReplica()` (util) uses `CreatorNode.getClockValue()` from libs
- `getNewOrExistingSyncReq()` (util) uses sync request dededuplicator from state machine (needs state of other sync requests on all threads in order to dedupe)
- config (it's trivial to remove this dependency by just reading the config values before starting the job and then passing them)

**issueSyncRequest job processor:**
- `getNewOrExistingSyncReq()` (util) uses sync request dededuplicator from state machine (needs state of other sync requests on all threads in order to dedupe)
- config (it's trivial to remove this dependency by just reading the config values before starting the job and then passing them)
- `models` might be OK but it might make new SQL connections if it's in its own thread
- `retrieveClockValueForUserFromReplica()` (util) uses `CreatorNode.getClockValue()` from libs
- SecondarySyncHealthTracker (util -- doesn't have any state so probably already OK)

**findSyncRequests job processor:** config, CNodeToSpIdMapManager, getNewOrExistingSyncReq (already mentioned limitations of each of these above)

**findReplicaSetUpdates job processor:**
- `config` and `CNodeToSpIdMapManager` (already mentioned limitations of each of these above)
- `CNodeHealthManager` (state machine util) needs state of each Content Node's health over time, so it's hard to make this pure

**monitorState job processor:** `config` and `CNodeHealthManager` (already mentioned limitations of each of these above)

Reply:
thanks for writing this all out, really helpful to see it this way

where possible we should attempt to re-design jobs to be more parallelizable, relying more on hooks / serializable (object) input data

for now its fine since this re-factor is independent of this parallelization work but would be good to reduce stateful dependencies where possible, move utils out of classes (i'm largely responsible for this haha) so they can be directly consumed

for libs, we should check if we actually need the stateful libs instance, and if we do not - how long the libs re-init step actually takes. my guess is its not that long and we can probably just stomach it on each job where required

i chatted with sebastian about longer term moving libs into a less stateful set of utilities that would be much more concurrency friendly, but for now we'll just work with what we have

the worst case (not really that bad) for libs is if there is a stateless function, we can just move it to a libs util and import that directly

Reply:
thanks for chatting with Sebastian and thinking about this some more. I documented this discussion [here](https://linear.app/audius/issue/CON-198/remove-stateful-dependencies-from-state-machine-jobs) so we'll make sure this work doesn't slip through the cracks